File: src/onion/compiler/env/ClosureLocalBinding.java
/* ************************************************************** *
 *                                                                *
 * Copyright (c) 2005, Kota Mizushima, All rights reserved.       *
 *                                                                *
 *                                                                *
 * This software is distributed under the modified BSD License.   *
 * ************************************************************** */
package onion.compiler.env;

import onion.lang.core.type.TypeRef;

/**
 * @author Kota Mizushima
 * Date: 2005/06/28
 */
public class ClosureLocalBinding extends LocalBinding {
  private int frame;
  
  public ClosureLocalBinding(int frame, int index, TypeRef type) {
    super(index, type);
    this.frame = frame;
  }

  public void setFrame(int frameIndex) {
    this.frame = frameIndex;
  }

  public int getFrame() {
    return frame;
  }
  
  public boolean equals(Object object) {
    if(!(object instanceof ClosureLocalBinding)) return false;
    ClosureLocalBinding bind = (ClosureLocalBinding) object;
    if(frame != bind.frame) return false;
    if(getIndex() != bind.getIndex()) return false;
    if(getType() != bind.getType()) return false;
    return true;
  }
  
  public int hashCode() {
    return frame + getIndex() + getType().hashCode();
  }
}

File: src/main/scala/onion/compiler/ClosureLocalBinding.scala
package onion.compiler


/**
 * @author Kota Mizushima
 * @param frameIndex Frame level (0 = current, 1 = parent, etc.)
 * @param isBoxed Whether the variable is boxed (for closure capture of mutable vars)
 */
class ClosureLocalBinding(val frameIndex: Int, index: Int, `type`: TypedAST.Type, isMutable: Boolean, isBoxed: Boolean = false)
  extends LocalBinding(index, `type`, isMutable, isBoxed) {

  override def equals(other: Any): Boolean = {
    other match {
      case bind: ClosureLocalBinding =>
        frameIndex == bind.frameIndex &&
        index == bind.index &&
        (tp eq bind.tp) &&
        isMutable == bind.isMutable &&
        isBoxed == bind.isBoxed
      case _ =>
        false
    }
  }

  override def hashCode: Int = frameIndex + index + tp.hashCode + (if (isMutable) 1 else 0) + (if (isBoxed) 2 else 0)
}
